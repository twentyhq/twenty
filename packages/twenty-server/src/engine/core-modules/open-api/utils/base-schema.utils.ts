import { OpenAPIV3_1 } from 'openapi-types';

import { computeOpenApiPath } from 'src/engine/core-modules/open-api/utils/path.utils';

export const API_Version = 'v0.1';

export const baseSchema = (
  schemaName: 'core' | 'metadata',
  serverUrl: string,
): OpenAPIV3_1.Document => {
  return {
    openapi: '3.1.1',
    info: {
      title: 'InsurOS Api',
      description: `This is a **InsurOS REST/API** playground based on the **OpenAPI 3.1 specification**.`,
      termsOfService:
        'https://github.com/insuros/twenty?tab=coc-ov-file#readme',
      contact: {
        email: 'felix@twenty.com',
      },
      license: {
        name: 'AGPL-3.0',
        url: 'https://github.com/insuros/twenty?tab=License-1-ov-file#readme',
      },
      version: API_Version,
    },
    // Testing purposes
    servers: [
      {
        url: `${serverUrl}/rest/${schemaName !== 'core' ? schemaName : ''}`,
        description: 'Production Development',
      },
    ],
    components: {
      securitySchemes: {
        bearerAuth: {
          type: 'http',
          scheme: 'bearer',
          bearerFormat: 'JWT',
          description:
            'Enter the token with the `Bearer: ` prefix, e.g. "Bearer abcde12345".',
        },
      },
    },
    security: [
      {
        bearerAuth: [],
      },
    ],
    externalDocs: {
      description: 'Find out more about **InsurOS**',
      url: 'https://insuros.ca',
    },
    paths: { [`/open-api/${schemaName}`]: computeOpenApiPath(serverUrl) },
  };
};
