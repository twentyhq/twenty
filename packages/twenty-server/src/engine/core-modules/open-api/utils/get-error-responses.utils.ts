import { type OpenAPIV3_1 } from 'openapi-types';

export const get400ErrorResponses = (): OpenAPIV3_1.ResponseObject => {
  return {
    description: 'Bad Request',
    content: {
      'application/json': {
        schema: {
          type: 'object',
          properties: {
            statusCode: { type: 'number' },
            messages: { type: 'array', items: { type: 'string' } },
            error: { type: 'string' },
          },
          example: {
            statusCode: 400,
            message: 'error message',
            error: 'Bad Request',
          },
        },
      },
    },
  };
};

export const get401ErrorResponses = (): OpenAPIV3_1.ResponseObject => {
  return {
    description: 'Unauthorized',
    content: {
      'application/json': {
        schema: {
          type: 'object',
          properties: {
            statusCode: { type: 'number' },
            message: { type: 'string' },
            error: { type: 'string' },
          },
          example: {
            statusCode: 401,
            message: 'Token invalid.',
            error: 'Unauthorized',
          },
        },
      },
    },
  };
};
