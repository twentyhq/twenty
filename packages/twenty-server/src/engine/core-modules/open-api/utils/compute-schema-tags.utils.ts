import { OpenAPIV3_1 } from 'openapi-types';
import { capitalize } from 'twenty-shared/utils';

import { ObjectMetadataEntity } from 'src/engine/metadata-modules/object-metadata/object-metadata.entity';

export const computeSchemaTags = (
  items: ObjectMetadataEntity[],
): OpenAPIV3_1.TagObject[] => {
  const results = [{ name: 'General', description: 'General requests' }];

  items.forEach((item) => {
    results.push({
      name: item.namePlural,
      description: `Object \`${capitalize(item.namePlural)}\``,
    });
  });

  return results;
};
