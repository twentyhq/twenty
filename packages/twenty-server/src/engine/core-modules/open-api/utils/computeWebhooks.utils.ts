import { OpenAPIV3_1 } from 'openapi-types';
import { capitalize } from 'twenty-shared/utils';

import { DatabaseEventAction } from 'src/engine/api/graphql/graphql-query-runner/enums/database-event-action';
import { ObjectMetadataEntity } from 'src/engine/metadata-modules/object-metadata/object-metadata.entity';

export const computeWebhooks = (
  type: DatabaseEventAction,
  item: ObjectMetadataEntity,
): OpenAPIV3_1.PathItemObject => {
  const updatedFields = {
    type: 'array',
    items: {
      type: 'string',
    },
  };

  return {
    post: {
      tags: [item.nameSingular],
      security: [],
      parameters: [
        {
          in: 'header',
          name: 'X-Twenty-Webhook-Signature',
          schema: {
            type: 'string',
          },
          description:
            'HMAC SHA256 signature of the request payload using the webhook secret. To compute the signature:\n' +
            '1. Concatenate `X-Twenty-Webhook-Timestamp`, a colon (:), and the JSON string of the request payload.\n' +
            '2. Compute the HMAC SHA256 hash using the shared secret as the key.\n' +
            '3. Send the resulting hex digest as this header value.\n' +
            'Example (Node.js):\n```javascript\n' +
            'const crypto = require("crypto");\n' +
            'const timestamp = "1735066639761";\n' +
            'const payload = JSON.stringify({...});\n' +
            'const secret = "your-secret";\n' +
            'const stringToSign = `${timestamp}:${JSON.stringify(payload)}`;\n' +
            'const signature = crypto.createHmac("sha256", secret)\n  .update(stringToSign)\n  .digest("hex");\n```',
          required: false,
        },
        {
          in: 'header',
          name: 'X-Twenty-Webhook-Timestamp',
          schema: {
            type: 'string',
          },
          description:
            'Unix timestamp of when the webhook was sent. This timestamp is included in the HMAC signature generation to prevent replay attacks.',
          required: false,
        },
        {
          in: 'header',
          name: 'X-Twenty-Webhook-Nonce',
          schema: {
            type: 'string',
          },
          description:
            'Unique identifier for this webhook request to prevent replay attacks. Consumers should ensure this nonce is not reused.',
          required: false,
        },
      ],
      requestBody: {
        content: {
          'application/json': {
            schema: {
              type: 'object',
              properties: {
                targetUrl: {
                  type: 'string',
                  example: 'https://example.com/incomingWebhook',
                },
                eventName: {
                  type: 'string',
                  example: `${item.nameSingular}.${type}`,
                },
                objectMetadata: {
                  type: 'object',
                  properties: {
                    id: {
                      type: 'string',
                      example: '370985db-22d8-4463-8e5f-2271d30913bd',
                    },
                    nameSingular: {
                      type: 'string',
                      enum: [item.nameSingular],
                    },
                  },
                },
                workspaceId: {
                  type: 'string',
                  example: '872cfcf1-c79f-42bc-877d-5829f06eb3f9',
                },
                webhookId: {
                  type: 'string',
                  example: '90056586-1228-4e03-a507-70140aa85c05',
                },
                eventDate: {
                  type: 'string',
                  example: '2024-02-14T11:27:01.779Z',
                },
                record: {
                  $ref: `#/components/schemas/${capitalize(item.nameSingular)} for Response`,
                },
                ...(type === DatabaseEventAction.UPDATED && { updatedFields }),
              },
            },
          },
        },
      },
      responses: {
        '200': {
          description:
            'Return a 200 status to indicate that the data was received successfully',
        },
      },
    },
  } as OpenAPIV3_1.PathItemObject;
};
