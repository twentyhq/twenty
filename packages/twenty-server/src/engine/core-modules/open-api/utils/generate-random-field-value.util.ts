import { faker } from '@faker-js/faker';
import { FieldMetadataType } from 'twenty-shared/types';
import { assertUnreachable, isDefined } from 'twenty-shared/utils';
import { v4 } from 'uuid';

import { FieldMetadataDefaultValue } from 'src/engine/metadata-modules/field-metadata/interfaces/field-metadata-default-value.interface';

import { FieldMetadataEntity } from 'src/engine/metadata-modules/field-metadata/field-metadata.entity';

export const generateRandomFieldValue = ({
  field,
}: {
  field: FieldMetadataEntity;
}): FieldMetadataDefaultValue => {
  switch (field.type) {
    case FieldMetadataType.UUID: {
      return v4();
    }

    case FieldMetadataType.TEXT: {
      return faker.string.fromCharacters(field.name);
    }

    case FieldMetadataType.PHONES: {
      return {
        primaryPhoneNumber: '06 10 20 30 40',
        primaryPhoneCallingCode: '+33',
        primaryPhoneCountryCode: 'FR',
        additionalPhones: [],
      };
    }

    case FieldMetadataType.EMAILS: {
      return {
        primaryEmail: faker.internet.email().toLowerCase(),
        additionalEmails: null,
      };
    }

    case FieldMetadataType.DATE:
    case FieldMetadataType.DATE_TIME: {
      return faker.date.soon();
    }

    case FieldMetadataType.BOOLEAN: {
      return false;
    }

    case FieldMetadataType.NUMBER: {
      return faker.number.float({ min: 1, max: 1_000 });
    }

    case FieldMetadataType.NUMERIC: {
      return faker.number.int({ min: 1, max: 1_000 });
    }

    case FieldMetadataType.LINKS: {
      return {
        primaryLinkLabel: '',
        primaryLinkUrl: faker.internet.url(),
        additionalLinks: [],
      };
    }

    case FieldMetadataType.CURRENCY: {
      return {
        amountMicros: faker.number.int({ min: 100, max: 1_000 }) * 1_000_000,
        currencyCode: 'EUR',
      };
    }

    case FieldMetadataType.FULL_NAME: {
      return {
        firstName: faker.person.firstName(),
        lastName: faker.person.lastName(),
      };
    }

    case FieldMetadataType.RATING: {
      return 'RATING_5';
    }

    case FieldMetadataType.SELECT: {
      return isDefined(field.options[0].value) ? field.options[0].value : [];
    }

    case FieldMetadataType.MULTI_SELECT: {
      return isDefined(field.options[0].value) ? [field.options[0].value] : [];
    }

    case FieldMetadataType.RELATION:
    case FieldMetadataType.MORPH_RELATION: {
      return null;
    }

    case FieldMetadataType.POSITION: {
      return 1;
    }

    case FieldMetadataType.ADDRESS: {
      return {
        addressStreet1: faker.location.streetAddress(),
        addressStreet2: faker.location.secondaryAddress(),
        addressCity: faker.location.city(),
        addressState: faker.location.state(),
        addressCountry: faker.location.country(),
        addressPostcode: faker.location.zipCode(),
        addressLat: faker.location.latitude(),
        addressLng: faker.location.longitude(),
      };
    }

    case FieldMetadataType.RAW_JSON: {
      return {};
    }

    case FieldMetadataType.RICH_TEXT:
    case FieldMetadataType.RICH_TEXT_V2: {
      return '';
    }

    case FieldMetadataType.ACTOR: {
      return {
        source: 'MANUAL',
        context: {},
        name: faker.person.fullName(),
        workspaceMemberId: null,
      };
    }

    case FieldMetadataType.ARRAY: {
      return [];
    }

    case FieldMetadataType.TS_VECTOR: {
      throw new Error(
        `We should not generate fake version for ${field.type} field`,
      );
    }

    default: {
      assertUnreachable(field.type, `Unsupported field type '${field.type}'`);
    }
  }
};
