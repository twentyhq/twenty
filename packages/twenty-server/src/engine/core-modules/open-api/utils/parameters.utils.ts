import { type OpenAPIV3_1 } from 'openapi-types';

import { OrderByDirection } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { DEFAULT_CONJUNCTION } from 'src/engine/api/rest/core/query-builder/utils/filter-utils/add-default-conjunction.utils';
import { FilterComparators } from 'src/engine/api/rest/core/query-builder/utils/filter-utils/parse-base-filter.utils';
import { Conjunctions } from 'src/engine/api/rest/core/query-builder/utils/filter-utils/parse-filter.utils';
import { DEFAULT_ORDER_DIRECTION } from 'src/engine/api/rest/input-factories/order-by-input.factory';

export const computeLimitParameters = (
  fromMetadata = false,
): OpenAPIV3_1.ParameterObject => {
  return {
    name: 'limit',
    in: 'query',
    description: 'Limits the number of objects returned.',
    required: false,
    schema: {
      type: 'integer',
      minimum: 0,
      maximum: fromMetadata ? 1000 : 60,
      default: fromMetadata ? 1000 : 60,
    },
  };
};

export const computeOrderByParameters = (): OpenAPIV3_1.ParameterObject => {
  return {
    name: 'order_by',
    in: 'query',
    description: `Sorts objects returned.  
    Should have the following shape: **field_name_1,field_name_2[DIRECTION_2],...**  
    Available directions are **${Object.values(OrderByDirection).join(
      '**, **',
    )}**.  
    Default direction is **${DEFAULT_ORDER_DIRECTION}**`,
    required: false,
    schema: {
      type: 'string',
    },
    examples: {
      simple: {
        value: `createdAt`,
        summary: 'A simple order_by param',
      },
      complex: {
        value: `id[${OrderByDirection.AscNullsFirst}],createdAt[${OrderByDirection.DescNullsLast}]`,
        summary: 'A more complex order_by param',
      },
    },
  };
};

export const computeDepthParameters = (): OpenAPIV3_1.ParameterObject => {
  return {
    name: 'depth',
    in: 'query',
    description: `Determines the level of nested related objects to include in the response.  
    - 0: Returns only the primary object's information.  
    - 1: Returns the primary object along with its directly related objects (with no additional nesting for related objects).  
    - 2: Returns the primary object, its directly related objects, and the related objects of those related objects.`,
    required: false,
    schema: {
      type: 'integer',
      enum: [0, 1, 2],
      default: 1,
    },
  };
};

export const computeFilterParameters = (): OpenAPIV3_1.ParameterObject => {
  return {
    name: 'filter',
    in: 'query',
    description: `Filters objects returned.  
    Should have the following shape: **field_1[COMPARATOR]:value_1,field_2[COMPARATOR]:value_2...
    To filter on composite type fields use **field.subField[COMPARATOR]:value_1
    **
    Available comparators are **${Object.values(FilterComparators).join(
      '**, **',
    )}**.  
    You can create more complex filters using conjunctions **${Object.values(
      Conjunctions,
    ).join('**, **')}**.  
    Default root conjunction is **${DEFAULT_CONJUNCTION}**.  
    To filter **null** values use **field[is]:NULL** or **field[is]:NOT_NULL**  
    To filter using **boolean** values use **field[eq]:true** or **field[eq]:false**`,

    required: false,
    schema: {
      type: 'string',
    },
    examples: {
      simple: {
        value: 'createdAt[gte]:"2023-01-01"',
        description: 'A simple filter param',
      },
      simpleNested: {
        value: 'emails.primaryEmail[eq]:foo99@example.com',
        description: 'A simple composite type filter param',
      },
      complex: {
        value:
          'or(createdAt[gte]:"2024-01-01",createdAt[lte]:"2023-01-01",not(id[is]:NULL))',
        description: 'A more complex filter param',
      },
    },
  };
};

export const computeStartingAfterParameters =
  (): OpenAPIV3_1.ParameterObject => {
    return {
      name: 'starting_after',
      in: 'query',
      description:
        'Returns objects starting after a specific cursor. You can find cursors in **startCursor** and **endCursor** in **pageInfo** in response data',
      required: false,
      schema: {
        type: 'string',
      },
    };
  };

export const computeEndingBeforeParameters =
  (): OpenAPIV3_1.ParameterObject => {
    return {
      name: 'ending_before',
      in: 'query',
      description:
        'Returns objects ending before a specific cursor. You can find cursors in **startCursor** and **endCursor** in **pageInfo** in response data',
      required: false,
      schema: {
        type: 'string',
      },
    };
  };

export const computeIdPathParameter = (): OpenAPIV3_1.ParameterObject => {
  return {
    name: 'id',
    in: 'path',
    description: 'Object id.',
    required: true,
    schema: {
      type: 'string',
      format: 'uuid',
    },
  };
};
