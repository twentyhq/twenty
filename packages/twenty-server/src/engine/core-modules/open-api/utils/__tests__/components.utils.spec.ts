import { FieldMetadataType } from 'twenty-shared/types';
import { EachTestingContext } from 'twenty-shared/testing';

import { NumberDataType } from 'src/engine/metadata-modules/field-metadata/interfaces/field-metadata-settings.interface';

import { objectMetadataItemMock } from 'src/engine/api/__mocks__/object-metadata-item.mock';
import { computeSchemaComponents } from 'src/engine/core-modules/open-api/utils/components.utils';
import { FieldMetadataEntity } from 'src/engine/metadata-modules/field-metadata/field-metadata.entity';
import { ObjectMetadataEntity } from 'src/engine/metadata-modules/object-metadata/object-metadata.entity';

describe('computeSchemaComponents', () => {
  it('should compute schema components', () => {
    expect(
      computeSchemaComponents([
        objectMetadataItemMock,
      ] as ObjectMetadataEntity[]),
    ).toMatchInlineSnapshot(`
{
  "ObjectName": {
    "description": undefined,
    "properties": {
      "fieldActor": {
        "properties": {
          "source": {
            "enum": [
              "EMAIL",
              "CALENDAR",
              "WORKFLOW",
              "API",
              "IMPORT",
              "MANUAL",
              "SYSTEM",
              "WEBHOOK",
            ],
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldAddress": {
        "properties": {
          "addressCity": {
            "type": "string",
          },
          "addressCountry": {
            "type": "string",
          },
          "addressLat": {
            "type": "number",
          },
          "addressLng": {
            "type": "number",
          },
          "addressPostcode": {
            "type": "string",
          },
          "addressState": {
            "type": "string",
          },
          "addressStreet1": {
            "type": "string",
          },
          "addressStreet2": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldArray": {
        "items": {
          "type": "string",
        },
        "type": "array",
      },
      "fieldBoolean": {
        "type": "boolean",
      },
      "fieldCurrency": {
        "properties": {
          "amountMicros": {
            "type": "number",
          },
          "currencyCode": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldDate": {
        "format": "date",
        "type": "string",
      },
      "fieldDateTime": {
        "format": "date-time",
        "type": "string",
      },
      "fieldEmails": {
        "properties": {
          "additionalEmails": {
            "items": {
              "format": "email",
              "type": "string",
            },
            "type": "array",
          },
          "primaryEmail": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldFullName": {
        "properties": {
          "firstName": {
            "type": "string",
          },
          "lastName": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldLinks": {
        "properties": {
          "primaryLinkLabel": {
            "type": "string",
          },
          "primaryLinkUrl": {
            "type": "string",
          },
          "secondaryLinks": {
            "items": {
              "description": "A secondary link",
              "properties": {
                "label": {
                  "type": "string",
                },
                "url": {
                  "format": "uri",
                  "type": "string",
                },
              },
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "fieldMultiSelect": {
        "items": {
          "enum": [
            "OPTION_1",
            "OPTION_2",
          ],
          "type": "string",
        },
        "type": "array",
      },
      "fieldNumber": {
        "type": "integer",
      },
      "fieldNumeric": {
        "type": "number",
      },
      "fieldPhones": {
        "properties": {
          "additionalPhones": {
            "items": {
              "type": "string",
            },
            "type": "array",
          },
          "primaryPhoneCallingCode": {
            "type": "string",
          },
          "primaryPhoneCountryCode": {
            "type": "string",
          },
          "primaryPhoneNumber": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldPosition": {
        "type": "number",
      },
      "fieldRating": {
        "enum": [
          "RATING_1",
          "RATING_2",
        ],
        "type": "string",
      },
      "fieldRawJson": {
        "type": "object",
      },
      "fieldRichText": {
        "type": "string",
      },
      "fieldSelect": {
        "enum": [
          "OPTION_1",
          "OPTION_2",
        ],
        "type": "string",
      },
      "fieldText": {
        "type": "string",
      },
      "fieldUuid": {
        "format": "uuid",
        "type": "string",
      },
    },
    "required": [
      "fieldNumber",
    ],
    "type": "object",
  },
  "ObjectName for Response": {
    "description": undefined,
    "properties": {
      "fieldActor": {
        "properties": {
          "name": {
            "type": "string",
          },
          "source": {
            "enum": [
              "EMAIL",
              "CALENDAR",
              "WORKFLOW",
              "API",
              "IMPORT",
              "MANUAL",
              "SYSTEM",
              "WEBHOOK",
            ],
            "type": "string",
          },
          "workspaceMemberId": {
            "format": "uuid",
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldAddress": {
        "properties": {
          "addressCity": {
            "type": "string",
          },
          "addressCountry": {
            "type": "string",
          },
          "addressLat": {
            "type": "number",
          },
          "addressLng": {
            "type": "number",
          },
          "addressPostcode": {
            "type": "string",
          },
          "addressState": {
            "type": "string",
          },
          "addressStreet1": {
            "type": "string",
          },
          "addressStreet2": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldArray": {
        "items": {
          "type": "string",
        },
        "type": "array",
      },
      "fieldBoolean": {
        "type": "boolean",
      },
      "fieldCurrency": {
        "properties": {
          "amountMicros": {
            "type": "number",
          },
          "currencyCode": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldDate": {
        "format": "date",
        "type": "string",
      },
      "fieldDateTime": {
        "format": "date-time",
        "type": "string",
      },
      "fieldEmails": {
        "properties": {
          "additionalEmails": {
            "items": {
              "format": "email",
              "type": "string",
            },
            "type": "array",
          },
          "primaryEmail": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldFullName": {
        "properties": {
          "firstName": {
            "type": "string",
          },
          "lastName": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldLinks": {
        "properties": {
          "primaryLinkLabel": {
            "type": "string",
          },
          "primaryLinkUrl": {
            "type": "string",
          },
          "secondaryLinks": {
            "items": {
              "description": "A secondary link",
              "properties": {
                "label": {
                  "type": "string",
                },
                "url": {
                  "format": "uri",
                  "type": "string",
                },
              },
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "fieldMultiSelect": {
        "items": {
          "enum": [
            "OPTION_1",
            "OPTION_2",
          ],
          "type": "string",
        },
        "type": "array",
      },
      "fieldNumber": {
        "type": "integer",
      },
      "fieldNumeric": {
        "type": "number",
      },
      "fieldPhones": {
        "properties": {
          "additionalPhones": {
            "items": {
              "type": "string",
            },
            "type": "array",
          },
          "primaryPhoneCallingCode": {
            "type": "string",
          },
          "primaryPhoneCountryCode": {
            "type": "string",
          },
          "primaryPhoneNumber": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldPosition": {
        "type": "number",
      },
      "fieldRating": {
        "enum": [
          "RATING_1",
          "RATING_2",
        ],
        "type": "string",
      },
      "fieldRawJson": {
        "type": "object",
      },
      "fieldRelation": {
        "items": {
          "$ref": "#/components/schemas/ToObjectMetadataName for Response",
        },
        "type": "array",
      },
      "fieldRichText": {
        "type": "string",
      },
      "fieldSelect": {
        "enum": [
          "OPTION_1",
          "OPTION_2",
        ],
        "type": "string",
      },
      "fieldText": {
        "type": "string",
      },
      "fieldUuid": {
        "format": "uuid",
        "type": "string",
      },
    },
    "type": "object",
  },
  "ObjectName for Update": {
    "description": undefined,
    "properties": {
      "fieldActor": {
        "properties": {
          "source": {
            "enum": [
              "EMAIL",
              "CALENDAR",
              "WORKFLOW",
              "API",
              "IMPORT",
              "MANUAL",
              "SYSTEM",
              "WEBHOOK",
            ],
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldAddress": {
        "properties": {
          "addressCity": {
            "type": "string",
          },
          "addressCountry": {
            "type": "string",
          },
          "addressLat": {
            "type": "number",
          },
          "addressLng": {
            "type": "number",
          },
          "addressPostcode": {
            "type": "string",
          },
          "addressState": {
            "type": "string",
          },
          "addressStreet1": {
            "type": "string",
          },
          "addressStreet2": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldArray": {
        "items": {
          "type": "string",
        },
        "type": "array",
      },
      "fieldBoolean": {
        "type": "boolean",
      },
      "fieldCurrency": {
        "properties": {
          "amountMicros": {
            "type": "number",
          },
          "currencyCode": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldDate": {
        "format": "date",
        "type": "string",
      },
      "fieldDateTime": {
        "format": "date-time",
        "type": "string",
      },
      "fieldEmails": {
        "properties": {
          "additionalEmails": {
            "items": {
              "format": "email",
              "type": "string",
            },
            "type": "array",
          },
          "primaryEmail": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldFullName": {
        "properties": {
          "firstName": {
            "type": "string",
          },
          "lastName": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldLinks": {
        "properties": {
          "primaryLinkLabel": {
            "type": "string",
          },
          "primaryLinkUrl": {
            "type": "string",
          },
          "secondaryLinks": {
            "items": {
              "description": "A secondary link",
              "properties": {
                "label": {
                  "type": "string",
                },
                "url": {
                  "format": "uri",
                  "type": "string",
                },
              },
              "type": "object",
            },
            "type": "array",
          },
        },
        "type": "object",
      },
      "fieldMultiSelect": {
        "items": {
          "enum": [
            "OPTION_1",
            "OPTION_2",
          ],
          "type": "string",
        },
        "type": "array",
      },
      "fieldNumber": {
        "type": "integer",
      },
      "fieldNumeric": {
        "type": "number",
      },
      "fieldPhones": {
        "properties": {
          "additionalPhones": {
            "items": {
              "type": "string",
            },
            "type": "array",
          },
          "primaryPhoneCallingCode": {
            "type": "string",
          },
          "primaryPhoneCountryCode": {
            "type": "string",
          },
          "primaryPhoneNumber": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "fieldPosition": {
        "type": "number",
      },
      "fieldRating": {
        "enum": [
          "RATING_1",
          "RATING_2",
        ],
        "type": "string",
      },
      "fieldRawJson": {
        "type": "object",
      },
      "fieldRichText": {
        "type": "string",
      },
      "fieldSelect": {
        "enum": [
          "OPTION_1",
          "OPTION_2",
        ],
        "type": "string",
      },
      "fieldText": {
        "type": "string",
      },
      "fieldUuid": {
        "format": "uuid",
        "type": "string",
      },
    },
    "type": "object",
  },
}
`);
  });

  const testsCases: EachTestingContext<
    Pick<
      FieldMetadataEntity<FieldMetadataType.NUMBER>,
      'id' | 'name' | 'type' | 'isNullable' | 'defaultValue' | 'settings'
    >
  >[] = [
    {
      title: 'Integer dataType with decimals',
      context: {
        id: 'number1',
        name: 'number1',
        type: FieldMetadataType.NUMBER,
        isNullable: false,
        defaultValue: null,
        settings: { type: 'number', decimals: 1, dataType: NumberDataType.INT },
      },
    },
    {
      title: 'Float without decimals',
      context: {
        id: 'number2',
        name: 'number2',
        type: FieldMetadataType.NUMBER,
        isNullable: false,
        defaultValue: null,
        settings: { type: 'number', dataType: NumberDataType.FLOAT },
      },
    },
    {
      title: 'Integer with a 0 decimals',
      context: {
        id: 'number3',
        name: 'number3',
        type: FieldMetadataType.NUMBER,
        isNullable: false,
        defaultValue: null,
        settings: { type: 'number', decimals: 0, dataType: NumberDataType.INT },
      },
    },
  ];

  it.each(testsCases)('$title', ({ context: field }) => {
    expect(
      computeSchemaComponents([
        {
          targetTableName: 'testingObject',
          id: 'mockObjectId',
          nameSingular: 'objectName',
          namePlural: 'objectsName',
          //@ts-expect-error Passing partial FieldMetadataEntity array
          fields: [field],
        },
      ]),
    ).toMatchSnapshot();
  });
});
