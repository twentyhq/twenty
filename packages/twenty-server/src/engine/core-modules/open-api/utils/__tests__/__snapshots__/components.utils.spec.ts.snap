// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computeSchemaComponents Float without decimals 1`] = `
{
  "ObjectName": {
    "description": undefined,
    "properties": {
      "number2": {
        "type": "number",
      },
    },
    "required": [
      "number2",
    ],
    "type": "object",
  },
  "ObjectName for Response": {
    "description": undefined,
    "properties": {
      "number2": {
        "type": "number",
      },
    },
    "type": "object",
  },
  "ObjectName for Update": {
    "description": undefined,
    "properties": {
      "number2": {
        "type": "number",
      },
    },
    "type": "object",
  },
}
`;

exports[`computeSchemaComponents Integer dataType with decimals 1`] = `
{
  "ObjectName": {
    "description": undefined,
    "properties": {
      "number1": {
        "type": "number",
      },
    },
    "required": [
      "number1",
    ],
    "type": "object",
  },
  "ObjectName for Response": {
    "description": undefined,
    "properties": {
      "number1": {
        "type": "number",
      },
    },
    "type": "object",
  },
  "ObjectName for Update": {
    "description": undefined,
    "properties": {
      "number1": {
        "type": "number",
      },
    },
    "type": "object",
  },
}
`;

exports[`computeSchemaComponents Integer with a 0 decimals 1`] = `
{
  "ObjectName": {
    "description": undefined,
    "properties": {
      "number3": {
        "type": "integer",
      },
    },
    "required": [
      "number3",
    ],
    "type": "object",
  },
  "ObjectName for Response": {
    "description": undefined,
    "properties": {
      "number3": {
        "type": "integer",
      },
    },
    "type": "object",
  },
  "ObjectName for Update": {
    "description": undefined,
    "properties": {
      "number3": {
        "type": "integer",
      },
    },
    "type": "object",
  },
}
`;
